text/0002-numeric-values.md

```diff
@@ -0,0 +1,485 @@
+
+```
+var x: u32 = 512
+// `512` becomes a u32 based on the type of `x`
+```
+
+Literals that do not fit in the result type will cause a compile error.
+
+```
+var x: u32 = -2
+// compile error: cannot coerce `-2` to `u32`
+```
+
+You cannot use [unary `~`](#integer-arithmetic) on an integer literal, it must be given a type, such as:
+
+```blight
+~512 // compile error: cannot infer type of constant numeric literal
+     // hint: unary operator '~' needs a result type
+
+~512.to(u32) // ok
+```
+
+Similarly, runtime-mutable variables cannot be assigned without annotating their type.
+
+> [!NOTE]
+> There is not yet an RFC for variable declarations. We will assume `var name: type` is the syntax.
+
+```blight
+var x = 512 // compile error: cannot infer type of constant numeric literal
+
+var x: u32 = 512 // ok
+var x = 512.to(u32) // ok, less preferred approach
+```
+
+## Constant Expression Evaluation
+
+Expressions that contain only numeric literals must be evaluated at compile time.
+
+```blight
+var x: u32 = (2.5 + 2.5) / 5
+// valid as `1` can coerce to `u32`
+
+var x: u32 = 5 / 6
+// compile error: cannot coerce `0.8333333333333334` to `u32`
+```
+
+If undefined behavior is caused by a constant expression, it is a compile error.
+
+```blight
+var x: u32 = 5 / 0
+// compile error: division by zero
+```
+
+## Widening during arithmetic
+
+When integers of different sizes are used in arithmetic, the compiler will widen the smaller integer to the larger integer.
+
+For example, if a variable of type `u32` named `x` existed, and the following code was run:
+
+```blight
+x + 512
+```
+
+`512` is widened to a u32. If the rhs was larger than `u32.max`, then this would be a compile error. Explicit widening of `x` is needed.
+
+```
+x.to(u64) + u32.max + 123
+```
+
+## Changes to Identifiers
+
+The following variable names are banned as they refer to built-in types, as they would shadow the implicit global.
+
+- `/[uif][0-9]+/`
+
+It is [not 100% decided](#variable-naming) how this impacts identifiers in general. It should probably be the job of that proposal to decide this.
+
+## Compiling `.to(T)`
+
+As there are only number types, this function will do nothing in most cases. It may need to compile to `Number(x)` or `BigInt(x)` depending on the action taken.
+
+## Compiling bitwise operations
+
+Spec compliant bit-shifting is not trivial to implement, it involves preserving the size for larger integers. The simplest way to do this is to convert JS numbers into [BigInt](https://tc39.es/proposal-bigint) and use the BigInt operators. This is not ideal, as it is slower than using the JS operators directly. It is also not ideal because it is not possible to use the BigInt operators on `f64` values.
+
+## Compiling arithmetic safety checks
+
+Saftey checks can be implemented somewhat trivially using a single temporary variable. This can be left as a global. It is unfortunate that so much code, as safety checks off allow this to compile to 3 characters.
```

Comment:
What does this sentence mean, this isn't valid english.

Reply:
correction: It is unfortunate that so much code is needed to compile this with saftey checks, as **with them disabled it allows this to compile to 3 characters